### PR TITLE
Fix rest api json serialization of KapuaPayload

### LINKED ISSUE #1662: Improve KapuaDataMessage JAXB Annotations
JAXB Annotations currently implemented on `KapuaDataMessage` will unmarshal the following JSON object:

```
{
	"type": "kapuaDataMessage",
	"position": {
		"type": "kapuaPosition",
		"latitude": 0,
		"longitude": 0
	},
	"channel": {
		"type": "kapuaDataChannel",
		"semanticParts": ["W1", "EVENT"]
	},
	"payload": {
		"type": "kapuaDataPayload",
		"metrics": {
			"metric": [{
				"valueType": "integer",
				"value": "1",
				"name": "Ruota"
			}]
		}
	},
	"deviceId": "{{DEVICE_ID}}",
	"clientId": "{{CLIENT_ID}}"
}
```

Starting from here we can make a couple of improvements:
- We should get rid of the `type` properties: inner objects should be unmarshalled automatically given their shape, not needing an explicit type declaration
- We should make `metrics` itself an array: there's no point in having an object with a single property with an array as value

### PR DESCRIPTION
**Pull request by Coduz**
Brief description of the PR.
Applied a patch to fix the JSON serialization of the `metrics` field of the `kapuaPayload` in REST API

**Related Issue**
This PR fixes #1662 

**Description of the solution adopted**
Wrapped the `DataMessages` JAX-RS resource with `DataMessageJson`. The new resources has the mappings to intercept `application/json` media types and uses custom objects (`JsonDatastoreMessage` and `JsonKapuaDataMessage`) to use correct mappings and convert them to the `DataMessages` specific objects.

REST API outputs will be:
***XML***
```
    <payload xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="kapuaDataPayload">
        <metrics>
            <metric>
                <valueType>float</valueType>
                <value>5.0</value>
                <name>temperatureExternal</name>
            </metric>
            <metric>
                <valueType>float</valueType>
                <value>10.5</value>
                <name>temperatureInternal</name>
            </metric>
            <metric>
                <valueType>float</valueType>
                <value>30.0</value>
                <name>temperatureExhaust</name>
            </metric>
            <metric>
                <valueType>integer</valueType>
                <value>0</value>
                <name>errorCode</name>
            </metric>
        </metrics>
    </payload>
```
***JSON***
```
    "payload": {
        "metrics": [
            {
                "valueType": "string",
                "value": "5.0",
                "name": "temperatureExternal"
            },
            {
                "valueType": "string",
                "value": "20.0",
                "name": "temperatureInternal"
            },
            {
                "valueType": "string",
                "value": "31.0",
                "name": "temperatureExhaust"
            },
            {
                "valueType": "string",
                "value": "0",
                "name": "errorCode"
            }
        ]
    }
```

**Screenshots**
_None_

**Any side note on the changes made**
_None_

### CODE REVIEW

**Comment by lorthirk** [comment]
We have to remember that with this current solution, Swagger only shows the JSON methods, not the XML ones since they are both annotated with the same name.

File: message/api/src/main/java/org/eclipse/kapua/message/xml/MetricsXmlAdapter.java [hunk 1]
```diff
@@ -40,6 +40,10 @@ public XmlAdaptedMetrics marshal(Map<String, Object> sourceMap) throws Exception
             }
         }
 
+        XmlUtil.marshal(new XmlAdaptedMetrics(adaptedMapItems));
```
Comment (stefanomorson) [inline message/api/src/main/java/org/eclipse/kapua/message/xml/MetricsXmlAdapter.java]: Remove this piece of code

File: rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/DataMessages.java [hunk 2]
```diff
@@ -51,52 +42,47 @@
...
 
 @Api(value = "Data Messages", authorizations = { @Authorization(value = "kapuaAccessToken") })
 @Path("{scopeId}/data/messages")
 public class DataMessages extends AbstractKapuaResource {
 
-    private static final KapuaLocator LOCATOR = KapuaLocator.getInstance();
-    private static final MessageStoreService MESSAGE_STORE_SERVICE = LOCATOR.getService(MessageStoreService.class);
-    private static final DatastoreObjectFactory DATASTORE_OBJECT_FACTORY = LOCATOR.getFactory(DatastoreObjectFactory.class);
-    private static final StorablePredicateFactory STORABLE_PREDICATE_FACTORY = LOCATOR.getFactory(StorablePredicateFactory.class);
+    protected static final KapuaLocator LOCATOR = KapuaLocator.getInstance();
```
Comment (stefanomorson) [inline rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/DataMessages.java]: Change to private

File: rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/DataMessagesJson.java [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                startDateParam,
+                endDateParam,
+                metricName,
+                metricType,
+                metricMinValue,
+                metricMaxValue,
+                offset,
+                limit);
+
+        List<JsonDatastoreMessage> asd = new ArrayList<>();
```
Comment (stefanomorson) [inline rest-api/resources/src/main/java/org/eclipse/kapua/app/api/resources/v1/resources/DataMessagesJson.java]: Change name of asd

**Comment by Coduz** [comment]
Hi all,

for some reasons Travis CI on PR build fails with this error:

> [INFO] ------------------------------------------------------------------------
> [INFO] Building kapua-rest-api-web 1.0.0-SNAPSHOT
> [INFO] ------------------------------------------------------------------------
> [INFO] 
> [INFO] --- maven-checkstyle-plugin:2.17:check (checkstyle-check) @ kapua-rest-api-web ---
> [INFO] 
> [INFO] --- jacoco-maven-plugin:0.7.9:prepare-agent (default-prepare-agent) @ kapua-rest-api-web ---
> [INFO] argLine set to -javaagent:/home/travis/.m2/repository/org/jacoco/org.jacoco.agent/0.7.9/org.jacoco.agent-0.7.9-runtime.jar=destfile=/home/travis/build/eclipse/kapua/rest-api/web/target/jacoco.exec
> [INFO] 
> [INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ kapua-rest-api-web ---
> [INFO] Using 'UTF-8' encoding to copy filtered resources.
> [INFO] Copying 4 resources
> [INFO] 
> [INFO] --- maven-resources-plugin:3.0.2:copy-resources (process-about.html) @ kapua-rest-api-web ---
> [INFO] Using 'UTF-8' encoding to copy filtered resources.
> [INFO] Copying 1 resource
> [INFO] 
> [INFO] --- maven-compiler-plugin:3.6.1:compile (default-compile) @ kapua-rest-api-web ---
> [INFO] Changes detected - recompiling the module!
> [INFO] Compiling 5 source files to /home/travis/build/eclipse/kapua/rest-api/web/target/classes
> [INFO] 
> [INFO] --- maven-resources-plugin:3.0.2:testResources (default-testResources) @ kapua-rest-api-web ---
> [INFO] Using 'UTF-8' encoding to copy filtered resources.
> [INFO] Copying 1 resource
> [INFO] 
> [INFO] --- maven-compiler-plugin:3.6.1:testCompile (default-testCompile) @ kapua-rest-api-web ---
> [INFO] No sources to compile
> [INFO] 
> [INFO] --- maven-surefire-plugin:2.20:test (default-test) @ kapua-rest-api-web ---
> [INFO] Tests are skipped.
> [INFO] 
> [INFO] --- jacoco-maven-plugin:0.7.9:report (default-report) @ kapua-rest-api-web ---
> [INFO] Skipping JaCoCo execution due to missing execution data file.
> [INFO] 
> [INFO] --- maven-dependency-plugin:3.0.1:unpack (unpack-swagger-ui) @ kapua-rest-api-web ---
> [INFO] Configured Artifact: org.eclipse.kapua.external:swagger-ui:2.1.4-CQ-10792:jar
> [INFO] Downloading: https://oss.sonatype.org/content/repositories/releases/org/eclipse/kapua/external/swagger-ui/2.1.4-CQ-10792/swagger-ui-2.1.4-CQ-10792.pom
> [INFO] Downloading: https://repository.apache.org/releases/org/eclipse/kapua/external/swagger-ui/2.1.4-CQ-10792/swagger-ui-2.1.4-CQ-10792.pom
> [WARNING] The POM for org.eclipse.kapua.external:swagger-ui:jar:2.1.4-CQ-10792 is missing, no dependency information available
> [INFO] Downloading: https://oss.sonatype.org/content/repositories/releases/org/eclipse/kapua/external/swagger-ui/2.1.4-CQ-10792/swagger-ui-2.1.4-CQ-10792.jar
> [INFO] Downloading: https://repository.apache.org/releases/org/eclipse/kapua/external/swagger-ui/2.1.4-CQ-10792/swagger-ui-2.1.4-CQ-10792.jar
> [INFO] ------------------------------------------------------------------------
> [INFO] Reactor Summary:
> [INFO] 
> [INFO] kapua-build-tools .................................. SUCCESS [  1.144 s]
> [INFO] kapua .............................................. SUCCESS [ 17.679 s]
> [INFO] kapua-service ...................................... SUCCESS [  4.265 s]
> [INFO] kapua-liquibase .................................... SUCCESS [  2.371 s]
> [INFO] kapua-service-api .................................. SUCCESS [  1.163 s]
> [INFO] kapua-user ......................................... SUCCESS [  0.149 s]
> [INFO] kapua-user-api ..................................... SUCCESS [  0.310 s]
> [INFO] kapua-security ..................................... SUCCESS [  1.042 s]
> [INFO] kapua-security-authorization ....................... SUCCESS [  0.166 s]
> [INFO] kapua-security-authorization-api ................... SUCCESS [  0.741 s]
> [INFO] kapua-security-authentication ...................... SUCCESS [  0.123 s]
> [INFO] kapua-security-authentication-api .................. SUCCESS [  0.484 s]
> [INFO] kapua-account ...................................... SUCCESS [  0.281 s]
> [INFO] kapua-account-api .................................. SUCCESS [  0.254 s]
> [INFO] kapua-commons ...................................... SUCCESS [  3.688 s]
> [INFO] kapua-locator ...................................... SUCCESS [  0.166 s]
> [INFO] kapua-guice ........................................ SUCCESS [  0.584 s]
> [INFO] kapua-message ...................................... SUCCESS [  0.193 s]
> [INFO] kapua-message-api .................................. SUCCESS [  0.274 s]
> [INFO] kapua-device ....................................... SUCCESS [  1.031 s]
> [INFO] kapua-device-api ................................... SUCCESS [  0.266 s]
> [INFO] kapua-device-registry .............................. SUCCESS [  0.410 s]
> [INFO] kapua-device-registry-api .......................... SUCCESS [  0.337 s]
> [INFO] kapua-datastore .................................... SUCCESS [  1.317 s]
> [INFO] kapua-datastore-client-api ......................... SUCCESS [  0.386 s]
> [INFO] kapua-test ......................................... SUCCESS [  0.968 s]
> [INFO] kapua-account-internal ............................. SUCCESS [  0.543 s]
> [INFO] kapua-datastore-api ................................ SUCCESS [  0.337 s]
> [INFO] kapua-user-internal ................................ SUCCESS [  1.091 s]
> [INFO] kapua-sso .......................................... SUCCESS [  0.108 s]
> [INFO] kapua-sso-jwt ...................................... SUCCESS [  0.162 s]
> [INFO] kapua-security-registration ........................ SUCCESS [  0.123 s]
> [INFO] kapua-security-registration-api .................... SUCCESS [  0.200 s]
> [INFO] kapua-security-certificate ......................... SUCCESS [  0.145 s]
> [INFO] kapua-security-certificate-api ..................... SUCCESS [  0.264 s]
> [INFO] kapua-security-shiro ............................... SUCCESS [  3.587 s]
> [INFO] kapua-message-internal ............................. SUCCESS [  0.682 s]
> [INFO] kapua-datastore-client-embedded .................... SUCCESS [  0.291 s]
> [INFO] kapua-datastore-client-transport ................... SUCCESS [  0.515 s]
> [INFO] kapua-datastore-client-rest ........................ SUCCESS [  0.456 s]
> [INFO] kapua-transport .................................... SUCCESS [  0.110 s]
> [INFO] kapua-transport-api ................................ SUCCESS [  0.190 s]
> [INFO] kapua-device-call .................................. SUCCESS [  0.261 s]
> [INFO] kapua-device-call-api .............................. SUCCESS [  0.337 s]
> [INFO] kapua-tag .......................................... SUCCESS [  0.087 s]
> [INFO] kapua-tag-api ...................................... SUCCESS [  0.130 s]
> [INFO] kapua-device-registry-internal ..................... SUCCESS [  1.039 s]
> [INFO] kapua-datastore-internal ........................... SUCCESS [  1.717 s]
> [INFO] kapua-device-asset ................................. SUCCESS [  0.225 s]
> [INFO] kapua-device-asset-api ............................. SUCCESS [  0.197 s]
> [INFO] kapua-translator ................................... SUCCESS [  0.329 s]
> [INFO] kapua-translator-api ............................... SUCCESS [  0.182 s]
> [INFO] kapua-device-commons ............................... SUCCESS [  0.368 s]
> [INFO] kapua-device-asset-internal ........................ SUCCESS [  0.294 s]
> [INFO] kapua-job .......................................... SUCCESS [  0.305 s]
> [INFO] kapua-job-api ...................................... SUCCESS [  0.320 s]
> [INFO] kapua-job-engine ................................... SUCCESS [  0.144 s]
> [INFO] kapua-job-engine-commons ........................... SUCCESS [  0.249 s]
> [INFO] kapua-device-asset-job ............................. SUCCESS [  0.243 s]
> [INFO] kapua-device-call-kura ............................. SUCCESS [  1.237 s]
> [INFO] kapua-device-configuration ......................... SUCCESS [  0.118 s]
> [INFO] kapua-device-configuration-api ..................... SUCCESS [  0.174 s]
> [INFO] kapua-device-configuration-internal ................ SUCCESS [  0.322 s]
> [INFO] kapua-device-configuration-job ..................... SUCCESS [  0.178 s]
> [INFO] kapua-device-packages .............................. SUCCESS [  0.185 s]
> [INFO] kapua-device-packages-api .......................... SUCCESS [  0.244 s]
> [INFO] kapua-device-packages-internal ..................... SUCCESS [  0.349 s]
> [INFO] kapua-device-packages-job .......................... SUCCESS [  0.322 s]
> [INFO] kapua-device-bundle ................................ SUCCESS [  0.133 s]
> [INFO] kapua-device-bundle-api ............................ SUCCESS [  0.173 s]
> [INFO] kapua-device-bundle-internal ....................... SUCCESS [  0.374 s]
> [INFO] kapua-device-bundle-job ............................ SUCCESS [  0.224 s]
> [INFO] kapua-device-command ............................... SUCCESS [  0.085 s]
> [INFO] kapua-device-command-api ........................... SUCCESS [  0.123 s]
> [INFO] kapua-device-command-internal ...................... SUCCESS [  0.218 s]
> [INFO] kapua-device-command-job ........................... SUCCESS [  0.167 s]
> [INFO] kapua-device-request ............................... SUCCESS [  0.080 s]
> [INFO] kapua-device-request-api ........................... SUCCESS [  0.141 s]
> [INFO] kapua-device-request-internal ...................... SUCCESS [  0.175 s]
> [INFO] kapua-endpoint ..................................... SUCCESS [  0.206 s]
> [INFO] kapua-endpoint-api ................................. SUCCESS [  0.141 s]
> [INFO] kapua-endpoint-internal ............................ SUCCESS [  0.248 s]
> [INFO] kapua-job-engine-api ............................... SUCCESS [  0.087 s]
> [INFO] kapua-job-engine-jbatch ............................ SUCCESS [  0.390 s]
> [INFO] kapua-job-internal ................................. SUCCESS [  1.044 s]
> [INFO] kapua-scheduler .................................... SUCCESS [  0.085 s]
> [INFO] kapua-scheduler-api ................................ SUCCESS [  0.138 s]
> [INFO] kapua-scheduler-quartz ............................. SUCCESS [  0.276 s]
> [INFO] kapua-security-certificate-internal ................ SUCCESS [  0.439 s]
> [INFO] kapua-broker ....................................... SUCCESS [  0.051 s]
> [INFO] kapua-broker-api ................................... SUCCESS [  0.084 s]
> [INFO] kapua-translator-kapua ............................. SUCCESS [  0.194 s]
> [INFO] kapua-translator-kapua-kura ........................ SUCCESS [  0.839 s]
> [INFO] kapua-transport-jms ................................ SUCCESS [  0.127 s]
> [INFO] kapua-translator-kura .............................. SUCCESS [  0.071 s]
> [INFO] kapua-translator-kura-jms .......................... SUCCESS [  0.161 s]
> [INFO] kapua-broker-core .................................. SUCCESS [  1.758 s]
> [INFO] kapua-security-registration-simple ................. SUCCESS [  0.308 s]
> [INFO] kapua-stream ....................................... SUCCESS [  0.062 s]
> [INFO] kapua-stream-api ................................... SUCCESS [  0.141 s]
> [INFO] kapua-stream-internal .............................. SUCCESS [  0.147 s]
> [INFO] kapua-tag-internal ................................. SUCCESS [  0.196 s]
> [INFO] kapua-transport-mqtt ............................... SUCCESS [  0.239 s]
> [INFO] kapua-translator-kura-mqtt ......................... SUCCESS [  0.314 s]
> [INFO] kapua-simulator-kura ............................... SUCCESS [  3.964 s]
> [INFO] kapua-qa-steps ..................................... SUCCESS [  3.332 s]
> [INFO] kapua-qa ........................................... SUCCESS [  0.600 s]
> [INFO] kapua-qa-openshift ................................. SUCCESS [  0.078 s]
> [INFO] kapua-sso-api ...................................... SUCCESS [  0.106 s]
> [INFO] kapua-sso-provider ................................. SUCCESS [  0.173 s]
> [INFO] kapua-sso-provider-generic ......................... SUCCESS [  0.188 s]
> [INFO] kapua-sso-provider-keycloak ........................ SUCCESS [  0.426 s]
> [INFO] kapua-extras ....................................... SUCCESS [  0.058 s]
> [INFO] kapua-foreignkeys .................................. SUCCESS [  0.087 s]
> [INFO] kapua-rest-api ..................................... SUCCESS [  0.384 s]
> [INFO] kapua-rest-api-core ................................ SUCCESS [  0.310 s]
> [INFO] kapua-rest-api-resources ........................... SUCCESS [  0.846 s]
> [INFO] kapua-rest-api-web ................................. FAILURE [  1.725 s]
> [INFO] kapua-client ....................................... SKIPPED
> [INFO] kapua-client-gateway ............................... SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: API ............. SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: SPI ............. SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: Provider ........ SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: Provider :: Core MQTT SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: Provider :: FUSE MQTT SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: Provider :: Eclipse Paho SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: Profile :: Eclipse Kura SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: Features ........ SKIPPED
> [INFO] Eclipse Kapua :: Gateway Client :: Features :: Karaf SKIPPED
> [INFO] kapua-assembly ..................................... SKIPPED
> [INFO] kapua-assembly-java-base ........................... SKIPPED
> [INFO] kapua-assembly-jetty-base .......................... SKIPPED
> [INFO] kapua-assembly-sql ................................. SKIPPED
> [INFO] kapua-assembly-broker .............................. SKIPPED
> [INFO] kapua-events-broker ................................ SKIPPED
> [INFO] kapua-assembly-api ................................. SKIPPED
> [INFO] kapua-dev-tools .................................... SKIPPED
> [INFO] kapua-cucumber-reports ............................. SKIPPED
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time: 01:25 min
> [INFO] Finished at: 2018-06-14T07:27:05+00:00
> [INFO] Final Memory: 173M/927M
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.0.1:unpack (unpack-swagger-ui) on project kapua-rest-api-web: Unable to find/resolve artifact. Failure to find org.eclipse.kapua.external:swagger-ui:jar:2.1.4-CQ-10792 in http://repo.maven.apache.org/maven2 was cached in the local repository, resolution will not be reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]
> [ERROR] 
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR] 
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
> [ERROR] 
> [ERROR] After correcting the problems, you can resume the build with the command
> [ERROR]   mvn <goals> -rf :kapua-rest-api-web

While on local, Travis CI Branch and Eclipse Jenkins build works fine.

I will merge anyway.

Regards, 

_Alberto_

**Pull request merged**